File: ios/FullscreenHandlerModule.swift
import BitmovinPlayer
import ExpoModulesCore

/**
 * Expo module for FullscreenHandler management with bidirectional communication.
 * Handles synchronous fullscreen state changes between native code and JavaScript.
 */
public class FullscreenHandlerModule: Module {
    /// In-memory mapping from `nativeId`s to `FullscreenHandlerBridge` instances.
    private var fullscreenHandlers: Registry<FullscreenHandlerBridge> = [:]

    /// ResultWaiter used for blocking thread while waiting for fullscreen state change
    private let waiter = ResultWaiter<Bool>()

    private var fullscreenActiveUpdateBuffer = LockedBox(value: [NativeId: Bool]())

    public func definition() -> ModuleDefinition {
        Name("FullscreenHandlerModule")

        OnDestroy {
            fullscreenHandlers.removeAll()
            waiter.removeAll()
            fullscreenActiveUpdateBuffer.update { value in
                value.removeAll()
            }
        }

        Events("onEnterFullscreen", "onExitFullscreen")

        AsyncFunction("registerHandler") { (nativeId: NativeId) in
            DispatchQueue.main.async { [weak self] in
                guard let self, fullscreenHandlers[nativeId] == nil else {
                    return
                }
                let handler = FullscreenHandlerBridge(
                    nativeId,
                    moduleRegistry: appContext?.moduleRegistry
                )
                if let isFullscreen = fullscreenActiveUpdateBuffer.value[nativeId] {
                    // Apply buffered value
                    handler.isFullscreenValueBox.update(isFullscreen)
                    fullscreenActiveUpdateBuffer.update { value in
                        value.removeValue(forKey: nativeId)
                    }
                }
                fullscreenHandlers[nativeId] = handler
            }
        }.runOnQueue(.main)

        AsyncFunction("destroy") { [weak self] (nativeId: NativeId) in
            self?.fullscreenHandlers.removeValue(forKey: nativeId)
            self?.fullscreenActiveUpdateBuffer.update { value in
                value.removeValue(forKey: nativeId)
            }
        }.runOnQueue(.main)

        AsyncFunction("notifyFullscreenChanged") { [weak self] (id: Int, isFullscreenEnabled: Bool) in
            self?.waiter.complete(id: id, with: isFullscreenEnabled)
        }

        AsyncFunction("setIsFullscreenActive") { [weak self] (nativeId: NativeId, isFullscreen: Bool) in
            guard let handler = self?.fullscreenHandlers[nativeId] else {
                // Buffer the value until the handler is registered
                self?.fullscreenActiveUpdateBuffer.update { value in
                    value[nativeId] = isFullscreen
                }
                return
            }
            handler.isFullscreenValueBox.update(isFullscreen)
        }.runOnQueue(.main)
    }

    func retrieve(_ nativeId: NativeId) -> FullscreenHandlerBridge? {
        fullscreenHandlers[nativeId]
    }

    /**
     * Handles fullscreen enter request from native code.
     * Called by FullscreenHandlerBridge when fullscreen should be entered.
     */
    func onFullscreenRequested(nativeId: NativeId) {
        guard let handler = retrieve(nativeId) else {
            return
        }

        let (id, wait) = waiter.make(timeout: 0.25)

        // Send event to JavaScript
        sendEvent("onEnterFullscreen", [
            "nativeId": nativeId,
            "id": id
        ])

        guard let result = wait() else {
            return
        }
        handler.isFullscreenValueBox.update(result)
    }

    func onFullscreenExitRequested(nativeId: NativeId) {
        guard let handler = retrieve(nativeId) else {
            return
        }

        let (id, wait) = waiter.make(timeout: 0.25)

        // Send event to JavaScript
        sendEvent("onExitFullscreen", [
            "nativeId": nativeId,
            "id": id
        ])

        guard let result = wait() else {
            return
        }
        handler.isFullscreenValueBox.update(result)
    }
}

File: android/src/main/java/com/bitmovin/player/reactnative/FullscreenHandlerModule.kt
package com.bitmovin.player.reactnative

import com.bitmovin.player.reactnative.ui.FullscreenHandlerBridge
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo module for FullscreenHandler management with bidirectional communication.
 * Handles synchronous fullscreen state changes between native code and JavaScript.
 */
class FullscreenHandlerModule : Module() {
    /**
     * In-memory mapping from `nativeId`s to `FullscreenHandlerBridge` instances.
     */
    private val fullscreenHandlers: Registry<FullscreenHandlerBridge> = mutableMapOf()

    /**
     * ResultWaiter used for blocking thread while waiting for fullscreen state change
     */
    private val waiter = ResultWaiter<Boolean>()

    override fun definition() = ModuleDefinition {
        Name("FullscreenHandlerModule")

        Events("onEnterFullscreen", "onExitFullscreen")

        AsyncFunction("registerHandler") { nativeId: String ->
            if (fullscreenHandlers[nativeId] == null) {
                fullscreenHandlers[nativeId] = FullscreenHandlerBridge(nativeId, this@FullscreenHandlerModule)
            }
        }

        AsyncFunction("destroy") { nativeId: String ->
            fullscreenHandlers.remove(nativeId)
        }

        AsyncFunction("notifyFullscreenChanged") { id: Int, isFullscreenEnabled: Boolean ->
            waiter.complete(id, isFullscreenEnabled)
        }

        AsyncFunction("setIsFullscreenActive") { nativeId: String, isFullscreenActive: Boolean ->
            fullscreenHandlers[nativeId]?.isFullscreen = isFullscreenActive
        }
    }

    /**
     * Retrieves the FullscreenHandlerBridge instance for the given nativeId.
     * This method maintains the same static access pattern as the legacy module.
     */
    fun getInstance(nativeId: String?): FullscreenHandlerBridge? = fullscreenHandlers[nativeId]

    /**
     * Handles fullscreen enter request from native code.
     * Called by FullscreenHandlerBridge when fullscreen should be entered.
     */
    fun requestEnterFullscreen(nativeId: String) {
        val handler = getInstance(nativeId) ?: return

        val (id, wait) = waiter.make(250) // 250ms timeout

        // Send event to JavaScript
        sendEvent(
            "onEnterFullscreen",
            mapOf(
                "nativeId" to nativeId,
                "id" to id,
            ),
        )

        val result = wait() ?: return
        handler.isFullscreen = result
    }

    /**
     * Handles fullscreen exit request from native code.
     * Called by FullscreenHandlerBridge when fullscreen should be exited.
     */
    fun requestExitFullscreen(nativeId: String) {
        val handler = getInstance(nativeId) ?: return

        val (id, wait) = waiter.make(250) // 250ms timeout

        // Send event to JavaScript
        sendEvent(
            "onExitFullscreen",
            mapOf(
                "nativeId" to nativeId,
                "id" to id,
            ),
        )

        val result = wait() ?: return
        handler.isFullscreen = result
    }
}
